libraries/AP_Camera/AP_Camera_SoloGimbal.cpp

```diff
@@ -0,0 +1,110 @@
+
+    } else if (gopro_capture_mode == GOPRO_CAPTURE_MODE_VIDEO) {
+        if (gopro_is_recording) {
+            // GoPro is recording, so stop recording
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro Recording Stop");
+            mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_SHUTTER,gopro_shutter_stop);
+        } else {
+            // GoPro is not recording, so start recording
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro Recording Start");
+            mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_SHUTTER,gopro_shutter_start);
+        }
+    } else {
+        gcs().send_text(MAV_SEVERITY_ERROR, "GoPro Unsupported Capture Mode");    
+    }
+}
+
+// Cycle the GoPro capture mode
+// This is so ArduPilot can toggle the capture mode of the GoPro directly, probably with an RC Aux function.
+// This is primarily for Solo's gimbal being installed on a vehicle other than a Solo.  The usual GoPro controls 
+// thorugh the Solo app and Solo controller do not use this, as it is done offboard on the companion computer.
+void AP_Camera_SoloGimbal::gopro_capture_mode_toggle()
+{
+    uint8_t gopro_capture_mode_values[4] = { };
+    
+    if (gopro_status != GOPRO_HEARTBEAT_STATUS_CONNECTED) {
+        gcs().send_text(MAV_SEVERITY_ERROR, "GoPro Not Available");
+        return;
+    }
+
+    if (gopro_capture_mode == GOPRO_CAPTURE_MODE_PHOTO) {
+        // Change to video mode
+        gopro_capture_mode_values[0] =  GOPRO_CAPTURE_MODE_VIDEO;
+        mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_CAPTURE_MODE,gopro_capture_mode_values);
+        gcs().send_text(MAV_SEVERITY_INFO, "GoPro changing to mode video");
+    } else if (gopro_capture_mode == GOPRO_CAPTURE_MODE_VIDEO) {
+        if (gopro_is_recording) {
+            // GoPro is recording, cannot change modes
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro recording, can't change modes");
+        } else {
+            // Change to camera mode
+            gopro_capture_mode_values[0] = GOPRO_CAPTURE_MODE_PHOTO;
+            mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_CAPTURE_MODE,gopro_capture_mode_values);
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro changing to mode photo");
+        }
+    } else {
+        gcs().send_text(MAV_SEVERITY_ERROR, "Unsupported GoPro Capture Mode");    
+    }
+}
+
+// heartbeat from the Solo gimbal GoPro
+void AP_Camera_SoloGimbal::handle_gopro_heartbeat(mavlink_channel_t chan, const mavlink_message_t &msg)
+{
+    mavlink_gopro_heartbeat_t report_msg;
+    mavlink_msg_gopro_heartbeat_decode(&msg, &report_msg);
+    gopro_is_recording = report_msg.flags & GOPRO_FLAG_RECORDING;
+    heartbeat_channel = chan;
+
+    switch(report_msg.status) {
```

Comment:
There are several case values missing here.  By casting to the relevant enumeration, you can get the compiler to tell you what you're missing:

```
diff --git a/libraries/AP_Camera/AP_Camera_SoloGimbal.cpp b/libraries/AP_Camera/AP_Camera_SoloGimbal.cpp
index e82588ddf1..915c2d15e1 100644
--- a/libraries/AP_Camera/AP_Camera_SoloGimbal.cpp
+++ b/libraries/AP_Camera/AP_Camera_SoloGimbal.cpp
@@ -81,19 +81,18 @@ void AP_Camera_SoloGimbal::handle_gopro_heartbeat(mavlink_channel_t chan, const
     gopro_is_recording = report_msg.flags & GOPRO_FLAG_RECORDING;
     heartbeat_channel = chan;
 
-    switch(report_msg.status) {
+    switch((GOPRO_HEARTBEAT_STATUS)report_msg.status) {
         case GOPRO_HEARTBEAT_STATUS_DISCONNECTED:
         case GOPRO_HEARTBEAT_STATUS_INCOMPATIBLE:
         case GOPRO_HEARTBEAT_STATUS_ERROR:
         case GOPRO_HEARTBEAT_STATUS_CONNECTED:
             gopro_status = (GOPRO_HEARTBEAT_STATUS)report_msg.status;
             break;
-        default:
-            // bad value
+        case GOPRO_HEARTBEAT_STATUS_ENUM_END:
             break;
     }
 
-    switch(report_msg.capture_mode){
+    switch((GOPRO_CAPTURE_MODE)report_msg.capture_mode){
         case GOPRO_CAPTURE_MODE_VIDEO:
         case GOPRO_CAPTURE_MODE_PHOTO:
         case GOPRO_CAPTURE_MODE_BURST:
@@ -101,10 +100,10 @@ void AP_Camera_SoloGimbal::handle_gopro_heartbeat(mavlink_channel_t chan, const
         case GOPRO_CAPTURE_MODE_MULTI_SHOT:
         case GOPRO_CAPTURE_MODE_PLAYBACK:
         case GOPRO_CAPTURE_MODE_UNKNOWN:
+        case GOPRO_CAPTURE_MODE_SETUP:
             gopro_capture_mode = (GOPRO_CAPTURE_MODE)report_msg.capture_mode;
             break;
-        default:
-            //bad value
+        case GOPRO_CAPTURE_MODE_ENUM_END:
             break;
     }
 }
```

Reply:
All in now.